### PR TITLE
Allowing contract method calls with explicit block numbers

### PR DESCRIPTION
**Pull request by amendelzon**
- Added retrieval of a block by its number on `ExecutionBlockRetriever`
- Covered cases with unit tests

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 4 issues

1. ![CRITICAL][CRITICAL] [ExecutionBlockRetriever.java#L27](https://github.com/rsksmart/rskj/blob/0bea977ddd7774e977d2694e658c5e66796400e9/rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java#L27): Unused import - org.spongycastle.util.encoders.Hex. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [ExecutionBlockRetriever.java#L32](https://github.com/rsksmart/rskj/blob/0bea977ddd7774e977d2694e658c5e66796400e9/rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java#L32): Unused import - java.math.BigInteger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![MINOR][MINOR] [ExecutionBlockRetriever.java#L27](https://github.com/rsksmart/rskj/blob/0bea977ddd7774e977d2694e658c5e66796400e9/rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java#L27): Remove this unused import 'org.spongycastle.util.encoders.Hex'. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [ExecutionBlockRetriever.java#L32](https://github.com/rsksmart/rskj/blob/0bea977ddd7774e977d2694e658c5e66796400e9/rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java#L32): Remove this unused import 'java.math.BigInteger'. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by tinchou** [comment]
By the way I'm not seeing Sonar reports here, but I got this on my inbox:

```
CRITICAL ExecutionBlockRetriever.java#L27: Unused import - org.spongycastle.util.encoders.Hex.
CRITICAL ExecutionBlockRetriever.java#L32: Unused import - java.math.BigInteger.
```

File: rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java [hunk 1]
```diff
@@ -71,6 +72,21 @@ public Block getExecutionBlock(String bnOrId) {
             return cachedBlock;
         }
 
+        if (bnOrId.startsWith("0x")) {
+            long executionBlockNumber;
+            try {
+                executionBlockNumber = Long.parseLong(bnOrId.substring(2), 16);
+            } catch (NumberFormatException e) {
+                throw new JsonRpcInvalidParamException(String.format("Not a number: %s", bnOrId), e);
```
[the remark this answers is not in the document]
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java]: :+1:

File: rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java [hunk 2]
```diff
@@ -71,6 +72,21 @@ public Block getExecutionBlock(String bnOrId) {
...
+            }
+
+            Block executionBlock = blockchain.getBlockByNumber(executionBlockNumber);
```
[the remark this answers is not in the document]
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java]: No, it's an invalid block number. The block isn't found as a consequence. If you're querying the blockchain, you should know its height in advance.

File: rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java [hunk 3]
```diff
@@ -71,6 +72,21 @@ public Block getExecutionBlock(String bnOrId) {
...
+            if (executionBlock == null) {
+                throw new JsonRpcInvalidParamException(String.format("Invalid block number %d", executionBlockNumber));
+            }
+            return executionBlock;
+        }
+
         throw new JsonRpcUnimplementedMethodException("Method only supports 'latest' and 'pending' as parameters so far.");
```
[the remark this answers is not in the document]
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java]: :+1:

**Comment by rskops** [comment]
SonarQube analysis reported 18 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 2 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [ExecutionBlockRetriever.java#L26](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java#L26): Unused import - org.ethereum.rpc.exception.JsonRpcUnimplementedMethodException. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [ExecutionBlockRetriever.java#L32](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java#L32): Unused import - javax.swing.text.html.Option. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [ExecutionBlockRetriever.java#L34](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java#L34): Unused import - java.util.regex.Pattern. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![MAJOR][MAJOR] [Utils.java#L40](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L40): Add a private constructor to hide the implicit public one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [Utils.java#L79](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L79): Only the sign of the result should be examined. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2200)
1. ![MAJOR][MAJOR] [Utils.java#L96](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L96): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Utils.java#L97](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L97): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [Utils.java#L103](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L103): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [Utils.java#L117](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L117): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MAJOR][MAJOR] [Utils.java#L122](https://github.com/rsksmart/rskj/blob/7406e5c264424c3c2478f97e06444ff5ac9793e3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L122): Method org.ethereum.util.Utils.getAddressShortString(byte[]) creates local variable-based synchronized collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALSYC_LOCAL_SYNCHRONIZED_COLLECTION)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 15 issues
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 2 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [Utils.java#L40](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L40): Add a private constructor to hide the implicit public one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [Utils.java#L79](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L79): Only the sign of the result should be examined. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2200)
1. ![MAJOR][MAJOR] [Utils.java#L96](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L96): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Utils.java#L97](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L97): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [Utils.java#L103](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L103): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [Utils.java#L117](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L117): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MAJOR][MAJOR] [Utils.java#L122](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L122): Method org.ethereum.util.Utils.getAddressShortString(byte[]) creates local variable-based synchronized collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALSYC_LOCAL_SYNCHRONIZED_COLLECTION)
1. ![MAJOR][MAJOR] [Utils.java#L139](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L139): Move the "0" string literal on the left side of this string comparison. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1132)
1. ![MAJOR][MAJOR] [Utils.java#L162](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L162): Method org.ethereum.util.Utils.getHashListShort(List) passes constant String of length 1 to character overridden method [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AUCPM_USE_CHARACTER_PARAMETERIZED_METHOD)
1. ![MINOR][MINOR] [Utils.java#L74](https://github.com/rsksmart/rskj/blob/b700dc6c01e1332c0bf33f8ab4bc21546ed629f3/rskj-core/src/main/java/org/ethereum/util/Utils.java#L74): Rename this field "_1000_" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS3008)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java [hunk 4]
```diff
@@ -71,6 +72,27 @@ public Block getExecutionBlock(String bnOrId) {
             return cachedBlock;
         }
 
-        throw new JsonRpcUnimplementedMethodException("Method only supports 'latest' and 'pending' as parameters so far.");
+        // Is the block specifier either a hexadecimal or decimal number?
+        Optional<Long> executionBlockNumber = Optional.empty();
+
+        if (Utils.isHexadecimalString(bnOrId)) {
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java]: This looks good, but I would prefer using [`Long.decode()`](https://docs.oracle.com/javase/7/docs/api/java/lang/Long.html#decode(java.lang.String)). We're reimplementing two of the three `decode`'s decoding formats!

File: rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java [hunk 5]
```diff
@@ -72,21 +72,27 @@ public Block getExecutionBlock(String bnOrId) {
...
             if (executionBlock == null) {
-                throw new JsonRpcInvalidParamException(String.format("Invalid block number %d", executionBlockNumber));
+                throw new JsonRpcInvalidParamException(String.format("Invalid block number %d", executionBlockNumber.get()));
             }
             return executionBlock;
         }
 
-        throw new JsonRpcUnimplementedMethodException("Method only supports 'latest' and 'pending' as parameters so far.");
+        // If we got here, the specifier given is unsupported
+        throw new JsonRpcInvalidParamException(String.format(
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java]: nitpick: we should be using `LATEST_ID` and `PENDING_ID`

**Comment by rskops** [comment]
SonarQube analysis reported 17 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 2 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [Utils.java#L40](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L40): Add a private constructor to hide the implicit public one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1118)
1. ![CRITICAL][CRITICAL] [Utils.java#L74](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L74): Explicitly declare the visibility for "_1000_". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2039)
1. ![CRITICAL][CRITICAL] [Utils.java#L79](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L79): Only the sign of the result should be examined. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2200)
1. ![MAJOR][MAJOR] [Utils.java#L69](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L69): Use the Java 8 Date and Time API instead. (sonar.java.source not set. Assuming 8 or greater.) [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2143)
1. ![MAJOR][MAJOR] [Utils.java#L96](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L96): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Utils.java#L97](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L97): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [Utils.java#L103](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L103): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [Utils.java#L122](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L122): Method org.ethereum.util.Utils.getAddressShortString(byte[]) creates local variable-based synchronized collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALSYC_LOCAL_SYNCHRONIZED_COLLECTION)
1. ![MAJOR][MAJOR] [Utils.java#L139](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L139): Move the "0" string literal on the left side of this string comparison. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1132)
1. ![MAJOR][MAJOR] [Utils.java#L162](https://github.com/rsksmart/rskj/blob/19b227592cbbc75bda4a3491a60b2817b6559cde/rskj-core/src/main/java/org/ethereum/util/Utils.java#L162): Method org.ethereum.util.Utils.getHashListShort(List) passes constant String of length 1 to character overridden method [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AUCPM_USE_CHARACTER_PARAMETERIZED_METHOD)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java (hunk 4 above)
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java]: Sorry about not responding this before. True, `Long.decode` already allows for parsing decimal, hexadecimal and octal, and I think that's cool. Nevertheless, I think we should enforce strict control on the functions we use to parse data that comes directly from a user request. In this case, if for some reason we overlook changes to `Long.decode` at some point, we could end up permitting input that we wouldn't originally want to allow. Does that make sense at all?

**Comment by rskops** [comment]
SonarQube analysis reported 17 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 2 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [Utils.java#L40](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L40): Add a private constructor to hide the implicit public one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1118)
1. ![CRITICAL][CRITICAL] [Utils.java#L74](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L74): Explicitly declare the visibility for "_1000_". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2039)
1. ![CRITICAL][CRITICAL] [Utils.java#L79](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L79): Only the sign of the result should be examined. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2200)
1. ![MAJOR][MAJOR] [Utils.java#L69](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L69): Use the Java 8 Date and Time API instead. (sonar.java.source not set. Assuming 8 or greater.) [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2143)
1. ![MAJOR][MAJOR] [Utils.java#L96](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L96): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Utils.java#L97](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L97): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [Utils.java#L103](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L103): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [Utils.java#L122](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L122): Method org.ethereum.util.Utils.getAddressShortString(byte[]) creates local variable-based synchronized collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALSYC_LOCAL_SYNCHRONIZED_COLLECTION)
1. ![MAJOR][MAJOR] [Utils.java#L139](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L139): Move the "0" string literal on the left side of this string comparison. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1132)
1. ![MAJOR][MAJOR] [Utils.java#L162](https://github.com/rsksmart/rskj/blob/da0666347ebcff079f9fd3ec3d18d99831a44440/rskj-core/src/main/java/org/ethereum/util/Utils.java#L162): Method org.ethereum.util.Utils.getHashListShort(List) passes constant String of length 1 to character overridden method [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AUCPM_USE_CHARACTER_PARAMETERIZED_METHOD)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**